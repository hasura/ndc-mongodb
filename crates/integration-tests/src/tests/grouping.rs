use insta::assert_yaml_snapshot;
use ndc_test_helpers::{
    and, asc, binop, column_aggregate, column_count_aggregate, dimension_column, field, grouping, or, ordered_dimensions, query, query_request, star_count_aggregate, target, value
};

use crate::{connector::Connector, run_connector_query};

#[tokio::test]
async fn runs_single_column_aggregate_on_groups() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        run_connector_query(
            Connector::SampleMflix,
            query_request().collection("movies").query(
                query()
                    // The predicate avoids an error when encountering documents where `year` is
                    // a string instead of a number.
                    .predicate(or([
                        binop("_gt", target!("year"), value!(0)),
                        binop("_lte", target!("year"), value!(0)),
                    ]))
                    .order_by([asc!("_id")])
                    .limit(10)
                    .groups(
                        grouping()
                            .dimensions([dimension_column("year")])
                            .aggregates([
                                (
                                    "average_viewer_rating",
                                    column_aggregate("tomatoes.viewer.rating", "avg"),
                                ),
                                ("max_runtime", column_aggregate("runtime", "max")),
                            ])
                            .order_by(ordered_dimensions()),
                    ),
            ),
        )
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn counts_column_values_in_groups() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        run_connector_query(
            Connector::SampleMflix,
            query_request().collection("movies").query(
                query()
                    .predicate(and([
                        binop("_gt", target!("year"), value!(1920)),
                        binop("_lte", target!("year"), value!(1923)),
                    ]))
                    .groups(
                        grouping()
                            .dimensions([dimension_column("rated")])
                            .aggregates([
                                // The distinct count should be 3 or less because we filtered to only 3 years
                                column_count_aggregate!("year_distinct_count" => "year", distinct: true),
                                column_count_aggregate!("year_count" => "year", distinct: false),
                                star_count_aggregate!("count"),
                            ])
                            .order_by(ordered_dimensions()),
                    ),
            ),
        )
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn groups_by_multiple_dimensions() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        run_connector_query(
            Connector::SampleMflix,
            query_request().collection("movies").query(
                query()
                    .predicate(binop("_lt", target!("year"), value!(1950)))
                    .order_by([asc!("_id")])
                    .limit(10)
                    .groups(
                        grouping()
                            .dimensions([
                                dimension_column("year"),
                                dimension_column("languages"),
                                dimension_column("rated"),
                            ])
                            .aggregates([(
                                "average_viewer_rating",
                                column_aggregate("tomatoes.viewer.rating", "avg"),
                            )])
                            .order_by(ordered_dimensions()),
                    ),
            ),
        )
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn combines_aggregates_and_groups_in_one_query() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        run_connector_query(
            Connector::SampleMflix,
            query_request().collection("movies").query(
                query()
                    .predicate(binop("_gte", target!("year"), value!(2000)))
                    .order_by([asc!("_id")])
                    .limit(10)
                    .aggregates([(
                        "average_viewer_rating",
                        column_aggregate("tomatoes.viewer.rating", "avg")
                    )])
                    .groups(
                        grouping()
                            .dimensions([dimension_column("year"),])
                            .aggregates([(
                                "average_viewer_rating_by_year",
                                column_aggregate("tomatoes.viewer.rating", "avg"),
                            )])
                            .order_by(ordered_dimensions()),
                    ),
            ),
        )
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn combines_fields_and_groups_in_one_query() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        run_connector_query(
            Connector::SampleMflix,
            query_request().collection("movies").query(
                query()
                    // The predicate avoids an error when encountering documents where `year` is
                    // a string instead of a number.
                    .predicate(or([
                        binop("_gt", target!("year"), value!(0)),
                        binop("_lte", target!("year"), value!(0)),
                    ]))
                    .order_by([asc!("_id")])
                    .limit(3)
                    .fields([field!("title"), field!("year")])
                    .order_by([asc!("_id")])
                    .groups(
                        grouping()
                            .dimensions([dimension_column("year")])
                            .aggregates([(
                                "average_viewer_rating_by_year",
                                column_aggregate("tomatoes.viewer.rating", "avg"),
                            )])
                            .order_by(ordered_dimensions()),
                    )
            ),
        )
        .await?
    );
    Ok(())
}
