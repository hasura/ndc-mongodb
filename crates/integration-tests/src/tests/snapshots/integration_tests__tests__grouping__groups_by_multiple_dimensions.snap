---
source: crates/integration-tests/src/tests/grouping.rs
expression: "run_connector_query(Connector::SampleMflix,\nquery_request().collection(\"movies\").query(query().predicate(binop(\"_lt\",\ntarget!(\"year\"),\nvalue!(1920))).groups(grouping().dimensions([dimension_column(\"year\"),\ndimension_column(\"languages\"),\ndimension_column(\"rated\"),]).aggregates([(\"average_viewer_rating\",\ncolumn_aggregate(\"tomatoes.viewer.rating\",\n\"avg\"),)]).order_by(ordered_dimensions()),),),).await?"
---
- groups:
    - dimensions:
        - 1891
        - - English
        - NOT RATED
      aggregates:
        average_viewer_rating: ~
    - dimensions:
        - 1893
        - ~
        - UNRATED
      aggregates:
        average_viewer_rating: 3
    - dimensions:
        - 1894
        - - English
        - NOT RATED
      aggregates:
        average_viewer_rating: 3.8
    - dimensions:
        - 1896
        - ~
        - ~
      aggregates:
        average_viewer_rating: 4
    - dimensions:
        - 1903
        - - English
        - TV-G
      aggregates:
        average_viewer_rating: 3.7
    - dimensions:
        - 1909
        - - English
        - G
      aggregates:
        average_viewer_rating: 3.6
    - dimensions:
        - 1911
        - - English
        - ~
      aggregates:
        average_viewer_rating: 3.4
    - dimensions:
        - 1912
        - - English
        - UNRATED
      aggregates:
        average_viewer_rating: 3.7
    - dimensions:
        - 1913
        - - English
        - TV-PG
      aggregates:
        average_viewer_rating: 3
    - dimensions:
        - 1914
        - - English
        - ~
      aggregates:
        average_viewer_rating: 3.0666666666666664
    - dimensions:
        - 1915
        - ~
        - NOT RATED
      aggregates:
        average_viewer_rating: 3.2
    - dimensions:
        - 1915
        - - English
        - NOT RATED
      aggregates:
        average_viewer_rating: 3.2
    - dimensions:
        - 1915
        - - English
        - PASSED
      aggregates:
        average_viewer_rating: 3.7
    - dimensions:
        - 1915
        - - French
        - NOT RATED
      aggregates:
        average_viewer_rating: 3.8
    - dimensions:
        - 1916
        - ~
        - ~
      aggregates:
        average_viewer_rating: 0
    - dimensions:
        - 1916
        - ~
        - NOT RATED
      aggregates:
        average_viewer_rating: 3.8
    - dimensions:
        - 1916
        - - English
        - APPROVED
      aggregates:
        average_viewer_rating: 3.1
    - dimensions:
        - 1916
        - - English
        - UNRATED
      aggregates:
        average_viewer_rating: 3.2
    - dimensions:
        - 1917
        - - English
        - ~
      aggregates:
        average_viewer_rating: 3.9
    - dimensions:
        - 1917
        - - English
        - UNRATED
      aggregates:
        average_viewer_rating: 4.1
    - dimensions:
        - 1918
        - - English
        - ~
      aggregates:
        average_viewer_rating: 3.6
    - dimensions:
        - 1919
        - ~
        - NOT RATED
      aggregates:
        average_viewer_rating: 3.7
    - dimensions:
        - 1919
        - - English
        - TV-G
      aggregates:
        average_viewer_rating: 3.3
